pubgrep/pubgrep.py

```diff
@@ -0,0 +1,526 @@
+    )
+    parser.add_argument(
+        "--skip", action="store_true", help="Skip PubChem server testing"
+    )
+    parser.add_argument(
+        "--verbosity",
+        type=int,
+        default=0,
+        choices=[0, 1, 2, 3],
+        help="Verbosity level",
+        required=False,
+    )
+    parser.add_argument(
+        "--opt",
+        action="store_true",
+        help="Optimize the structure of the compound",
+        required=False,
+        default=False,
+    )
+    parser.add_argument(
+        "--basedir",
+        type=Path,
+        default="pubchem_compounds",
+        help="Base directory for the compound data",
+        required=False,
+    )
+    parser.add_argument(
+        "--hlgap_thr",
+        type=float,
+        default=0.5,
+        help="Threshold for the HOMO-LUMO gap",
+        required=False,
+    )
+    return parser.parse_args()
+
+
+def cli():
+    """
+    This function is used to parse the command line arguments.
+    """
+    try:
+        args = get_args()
+        verbosity = args.verbosity
+        compounds = args.compounds
+        input_format = args.input
+        output_format = args.output
+        optimization = args.opt
+        skip = args.skip
+        basedir = args.basedir
+        hlgap_thr = args.hlgap_thr
+
+        xtb_path = Path(shutil.which("xtb")).resolve()
+
+        pubgrep(
+            xtb_path=xtb_path,
+            basedir=basedir,
+            compounds=compounds,
+            input_format=input_format,
+            output_format=output_format,
+            optimization=optimization,
+            hlgap_thr=hlgap_thr,
+            skip=skip,
+            verbosity=verbosity,
+        )
+        sys.exit(0)
+    except Exception as e:
+        print(f"Error: {e}", file=sys.stderr)
+        sys.exit(1)
+
+
+def pubgrep(
+    xtb_path: Path,
+    basedir: Path,
+    compounds: list[str] | Path | str,
+    input_format: str,
+    output_format: str,
+    optimization: bool,
+    hlgap_thr: float,
+    skip: bool,
+    verbosity: int,
+):
+    """
+    This function is used to search the compounds in the PubChem database.
+    """
+
+    if output_format not in ["sdf", "logP", "list"]:
+        raise ValueError("Invalid output format.")
+
+    if not skip:
+        test_pubchem_server(verbosity)
+
+    # if len of compounds is 1, take it as a single compound
+    if isinstance(compounds, list):
+        if len(compounds) == 1:
+            compounds = compounds[0]
+
+    compound_list = []
+    if isinstance(compounds, list):
+        for compound in compounds:
+            compound_list.append(compound)
+    elif isinstance(compounds, Path) or Path(compounds).is_file():  # type: ignore
+        with open(compounds, "r", encoding="utf-8") as file:  # type: ignore
+            compound_list = [line.strip() for line in file.readlines()]
+    elif isinstance(compounds, str):
+        compound_list.append(compounds)
+
```

Comment:
Nitpick: Including an else block that just throws an error for unrecognized formats, helps avoiding wrong input types.